GVFS/GVFS.Common/GitHubUpgrader.cs

```diff
@@ -545,6 +568,43 @@ private bool AssetInstallerNameCompare(Asset asset, params string[] expectedFile
             this.tracer.RelatedEvent(EventLevel.Informational, message, metadata);
         }
 
+        private bool VerifyCodeSignature(string path, string certCN, out string error)

```

Comment:
Are we doing something similar with NuGet? Is there a way that we can generalize this method?

Reply:
Tried verifying a Nuget package with this method - I get the exception " System.Security.Cryptography.CryptographicException: Cannot find the requested object."

Tried using the SignTool to verify nuget package - that also didn't work.
```
C:\Users\alsha>signtool verify /pa C:\Users\alsha\Downloads\nunit.3.11.0.nupkg
File: C:\Users\alsha\Downloads\nunit.3.11.0.nupkg
Index  Algorithm  Timestamp
========================================
SignTool Error: This file format cannot be verified because it is not
        recognized.

Number of errors: 1
```
Couldn't yet find an API/Tool that could be generic enough (Even Windows Explorer does not show certificate info for signed Nuget packages.) Will check with@jamill (he is looking into verifying nuget packages) and update this thread.

Reply:
Please do check, as I expect this crypto stuff to be useful for both.

Reply:
The other benefit to splitting it out preemptively is that you can inject a mock during unit tests to ensure the GitHubUpgrader responds correctly to incorrect certificate results.

Reply:
>> Please do check, as I expect this crypto stuff to be useful for both.

Did some more digging. Found a few issues with using C# API's - 1. it doesn't validate Nuget packages. 2. it doesn't help validate that the signed binary has not been tampered with (it  only checks that the certificate used to sign the binary is valid). 3. C# API also don't work on the Mac. Couldn't see much value in using C# API (vs using more capable platform native).

Alternative option was to use platform specific APIs (WinTrustAPI on Windows and in future may be use `codesign` tool on Mac). There is a `PowerShell` wrapper for WinTrustAPI, `Get-AuthenticodeSignature()` which is easy to use. This method cannot be used to validate Nuget packages however, but it validates the signing certificate and checks binary tampering. `Nuget Upgrader` will have to use Nuget specific API to validate the container package, but once after the package has been unarchived to the `Downloads` directory, it can still use this method (`GVFSPlatform.Instance.TryVerifyAuthenticodeSignature()`) before launching the exes.